Add Archetype to CLI

TylerJewell opened the pull request:
### What does this PR do?
Adds a new command to the CLI, `archetype` which can be used to generate, build, and run a new custom assembly. It supports all of the different archetypes that are generated and gives the user the option to select which archetype to use. This utility uses the `eclipse/che-dev` Docker imae to execute maven including the 'mvn clean install' of the assembly itself.  The CLI reuses itself to run the build assembly.

#### Changelog
[cli] Add new 'archetype' command for generating, building and running custom assemblies
[cli] Remove orphaned nightly images if re-pulled

#### Release Notes
We have added a new command to the CLI, `archetype` which can be used to generate custom assemblies of Eclipse Che and Codenvy.

An assembly is a bundling of extensions, plugins, stacks, agents, branding elements, and a CLI
that can be built into a new binary for distribution. In essence, an assemly is a custom che. Custom assemblies are used by Che developers and customizers to create new products with Che running embedded.

An archetype is a maven technique for generating code templates. A single archetype has an ID and
generates a complete custom assembly. Differnent archetypes generate assemblies with different
types of customizations. We make each archetype customize the minimal number of features to make
learning about customizations simpler.

Your host system must have Maven 3.3+ installed to facilitate generation and compiling of custom
assemblies. You must pass in your Maven's M2 repository path on your host. Our archetype generator
will download libraries into that repository making repeated compilations faster over time.
On most Linux based systems, your M2 is located at '/home/user/.m2' and it is '%USERPROFILE%/.m2'
for Windows. We default your M2 repository to '/home/user/.m2'. Use the '/m2' mount to chnage this.

Your custom assembly will be generated in the host path mounted to '/archetype'. This generates a
Maven multi-module project. You can enter the folder and build it with 'mvn clean install' or use
this utility to build it. Compiling an assembly requires other tools like Java, Angular, Go to be
installed on your host system. However, if you use this tool to compile your custom assembly we
use 'eclipse/che-dev' Docker image which contains all of these utilities preinstalled. It is simple
but is a large download >1GB and compilation is slower than using your host since the Docker
container is performing compilation against files that are host-mounted.

The syntax of the command is 
```
USAGE: DOCKER_PARAMS eclipse/che-cli:nightly archetype ACTION [PARAMETERS]

Use an archetype to generate, build or run a custom che assembly

MANDATORY DOCKER PARAMETERS:
  -v <path>:/archetype        Local path where your custom assembly will be generated

OPTIONAL DOCKER PARAMETERS:
  -v <path>:/m2               Local path to your host's Maven M2 repository

ACTION:
  all                         Generate, build and run a new custom assembly
  generate                    Generate a new custom assembly to folder mounted to '/archetype'
  build                       Uses 'eclipse/che-dev' image to compile archetype in '/archetype'
  run                         Starts che from custom assembly in '/archetype'

PARAMETERS:
  --archid=<id>               Different archetypes generate different types of customizations
  --archversion=<version>     Sets archetype version - default = tag of CLI image
  --version=<version>         Sets custom assembly version - default = archetype version
  --group=<group>             Sets groupId of generated assembly - default = com.sample
  --id=<id>                   Sets artifactId of generated assembly - default = assembly
  --no:interactive            Disables interactive mode
```

We provide a variety of different archetypes that will generate custom assemblies with customizations for a small portion of Che. This is provided to simplify the learning experience for new developers with Che. Custom assemblies will add code, change some configuration files, and also add new software dependencies. Developers can then use Maven (or this `archetype` utility) to compile those changes into a package that can be run with the CLI.

For example:
```
docker run -it --rm
  -v /var/run/docker.sock:/var/run/docker.sock 
  -v /c/archetype:/archetype 
  -v /c/tmp:/data 
  -v /c/Users/Tyler/.m2/repository:/m2 
    eclipse/che-cli:5.7.0 
      archetype all --no:interactive
```
This is syntax on Windows will:
1. Generate a custom assembly into the `C:\archetype` folder.
2. The custom assembly will accept all of the defaults without prompting the user. In this case, the menu archetype will be generated with a custom menu entry that launches a custom action.
3. The custom assembly's Maven dependencies will be saved to the host's maven repository located at `C:\Users\Tyler\.m2\repository`. These dependencies will be downloaded and compiled into the assembly during the build phase.
4. The CLI will launch another CLI instance to run the custom assembly with that custom assembly's data folder being `C:\tmp`. 

#### Linux
On Linux, the Maven utility launched by the CLI may have problems writing files to your host system during the generation phase. To overcome this limitation, in the `docker run ...` syntax before the image name add `--user={uid}` where UID is the user identity of your current users. This user identity will be passed into the various CLI utilities to execute under that user identity.

#### Docs PR
TBD

skabashnyuk commented:
1. ```-v $HOME/.che/data:/data -v /var/run/docker.sock:/var/run/docker.sock```  looks like this two options is mandatory
2. We should be consitent with  -v <path>:/m2
parameter it should be  $HOME/.m2 or $HOME/.m2/repo (means actual location of localRepository).
In some cases $HOME/.m2 can be a problem. If user override localRepository in settings.xml when we will have
trouble with this path in docker container.

3. ```ARCHETYPE_VERSION=5.4.0-SNAPSHOT```
4. Generate is not working on linux . I think user has no write permission  for che-build
```
user@271c2fe72c34:~/che-build$ ls
user@271c2fe72c34:~/che-build$ touch 1
touch: cannot touch ‘1’: Permission denied
user@271c2fe72c34:~/che-build$
user@271c2fe72c34:~/che-build$
```

5.```-DarchetypeVersion=$ARCHETYPE_VERSION -Dversion=$ARCHETYPE_VERSION```
 This is two different parameters we should distinct them



 Question: shall we stop development and testing of *.sh files inside of archetypes 
 https://github.com/eclipse/che-archetypes/tree/master/che-archetype-resources/src/main/resources

TylerJewell commented:
See below 

Tyler

> On Feb 20, 2017, at 4:16 AM, Sergii Kabashniuk <notifications@github.com> wrote:
> 
> -v $HOME/.che/data:/data -v /var/run/docker.sock:/var/run/docker.sock looks like this two options is mandatory
> 
Tyler - can you elaborate what and where you mean?
> 
> We should be consitent with -v :/m2
> parameter it should be $HOME/.m2 or $HOME/.m2/repo (means actual location of localRepository).
> In some cases $HOME/.m2 can be a problem. If user override localRepository in settings.xml when we will have
> trouble with this path in docker container.
> 
This creates some problems with syntax as having three mounts is challenging.  We will have to think about how we can simplify this.
> 
> ARCHETYPE_VERSION=5.4.0-SNAPSHOT
> 
> Generate is not working on linux . I think user has no write permission for che-build
> 
> user@271c2fe72c34:~/che-build$ ls
> user@271c2fe72c34:~/che-build$ touch 1
> touch: cannot touch ‘1’: Permission denied
> user@271c2fe72c34:~/che-build$
> user@271c2fe72c34:~/che-build$
Can you issue a commit to fix this?  There are examples of how to make the file system writable in the cli.  I don't have an easy way to test on Linux a fix but roman will know what to do.
> 5.-DarchetypeVersion=$ARCHETYPE_VERSION -Dversion=$ARCHETYPE_VERSION
> This is two different parameters we should distinct them
> 
I didn't want to have them separate as they are confusing separated :).  What we really need is a way for us to tell the user what version of the archetype is available.   The end user will have no ability to type this themselves as they do not follow our versioning closely.  So it needs a discovery mechanism.
> Question: shall we stop development and testing of *.sh files inside of archetypes
> https://github.com/eclipse/che-archetypes/tree/master/che-archetype-resources/src/main/resources
> 
We should still finish them.  So that people can run on the host.  On these, it should be run.sh not Che-run.sh.  We will let users pass --che or --codenvy as a parameter.
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

skabashnyuk commented:
> Tyler - can you elaborate what and where you mean?

```archetype  generate``` is not working without ```-v $HOME/.che/data:/data -v /var/run/docker.sock:/var/run/docker.sock``` 


> This creates some problems with syntax as having three mounts is challenging.  We will have to think about how we can simplify this.

What if we from start will use only path to repository folder?

>Can you issue a commit to fix this?

will talk with @riuvshin 

>I didn't want to have them separate as they are confusing separated :) 

Am I understand correctly DarchetypeVersion will be some how discovered, Dversion will be provided by user to have logical ability to set his project version != archetypeVersion

> We should still finish them. 

It's double our time, we need test sh and cli. Not sure it's optimal

TylerJewell commented:
Yes, `archetype:generate` still requires the same elements as the full CLI.  Both `-v /var/run/docker.sock:/var/run/docker.sock` and `:/data` mounts are always required for any command of the CLI including the `archetype` command.

I want the archetypeVersion and projectVersion to be identical when generating.  We can offer the end user the ability to go into their custom assembly and change versions manually.  But for the purposes of the CLI itself, we can make certain hard coded assumptions for simplification.  Once a user has a custom assembly then they can push that into a repository and then continue to use it.

Yeah, I suppose that if we can eventually get the CLI to work in every scenario, then we can probably stop maintaining the shell scripts. But it's not really 2x the effort as the logic is largely the same between the two.

skabashnyuk commented:
>I want the archetypeVersion and projectVersion to be identical when generating

I'm really upset with this statement. I  believe this assumption can simplify a couple of things but it will cause a lot of potential issues.

TylerJewell commented:
About the comment on M2 home vs. M2 repository.  I think we only really need access to the M2 repository.  We don't need access to M2 home since that is accounted for within the Docker image itself.  If so, then the :/m2 mount should be re-thought of as pointing directly to the M2 repository and then end users should mount their M2 repository path.  That does make it a lot harder to explain to end users.  

Those that are familiar with maven and customize it should be ok in finding the right location. However, if you are largely unfamiliar with Maven then it may not be obvious to you where the M2 repository is located.

I do not understand why any emotion exists on the versioning issue. The hardest part of the custom assembly will be first giving the end user a list of viable archetype versions that the end user can use.  If we set their first version in their pom to be the same value, then isn't there only a single location that they need to change to move their assembly version to a custom value?

TylerJewell commented:
@skabashnyuk - have done a couple things:
1. Added in a separation between archetype version and assembly version.  I am not a super fan of this, but understand the concern.
2. Added in some checks to verify that the assembly exists before executing a build or a run.
3. If docker for mac or docker for windows, applies the value of --user={uid} from the CLI and passes it into the generate or build command.  I also then mount the `/etc` folders that are required.

skabashnyuk commented:
test on linux and OSX
```
docker run -it --rm  \
   -v /var/run/docker.sock:/var/run/docker.sock  \
   -v $HOME/dev/tmp/arch:/archetype \
   -v $HOME/.m2/repository:/m2 \
   -v $HOME/.che/data:/data \
   eclipse/che:nightly \
    archetype  generate --fast \
   --archid=che-plugin-ide-menu-archetype \
   --archversion=5.4.0-SNAPSHOT \
   --group=ofef.copm \
   --version=1.0-SNAPSHOT \
   --id=com-wizard \
   --no:interactive
```
and have

```
Downloaded: https://repo.maven.apache.org/maven2/org/codehaus/groovy/groovy/1.8.3/groovy-1.8.3.jar (5394 KB at 426.5 KB/sec)
[INFO] Generating project in Batch mode
[WARNING] Archetype not found in any catalog. Falling back to central repository (http://repo.maven.apache.org/maven2).
[WARNING] Use -DarchetypeRepository=<your repository> if archetype's repository is elsewhere.
Downloading: http://repo.maven.apache.org/maven2/org/eclipse/che/archetype/che-plugin-ide-menu-archetype/5.4.0-SNAPSHOT/maven-metadata.xml
Downloading: http://repo.maven.apache.org/maven2/org/eclipse/che/archetype/che-plugin-ide-menu-archetype/5.4.0-SNAPSHOT/che-plugin-ide-menu-archetype-5.4.0-SNAPSHOT.pom
[WARNING] The POM for org.eclipse.che.archetype:che-plugin-ide-menu-archetype:jar:5.4.0-SNAPSHOT is missing, no dependency information available
Downloading: http://repo.maven.apache.org/maven2/org/eclipse/che/archetype/che-plugin-ide-menu-archetype/5.4.0-SNAPSHOT/che-plugin-ide-menu-archetype-5.4.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 50.668 s
[INFO] Finished at: 2017-02-21T11:00:48+00:00
```

skabashnyuk commented:
After https://github.com/eclipse/che/commit/9586e7d677cf0c3c877bc04ed6254de5e77c0f9d
generate is working on OSX and linux. on Linux project generated with root permissions

skabashnyuk commented:
@TylerJewell @garagatyi pointed me to the fact that if we mount maven repository and artifacts will be created under root. Use will blocked to create any files in created folder. He can be blocked to use maven on host machine. 

@benoitf @TylerJewell any ideas how to avoid creating files with root permission on host machine ?

benoitf commented:
@skabashnyuk using docker run -u <uid> and mounting passwd/group file should allow to write files not in root

skabashnyuk commented:
According to @TylerJewell documentation it should be like this 
```
docker run  -it --rm  \
   -v /var/run/docker.sock:/var/run/docker.sock  \
   -v $HOME/dev/tmp/arch:/archetype \
   -v $HOME/hdd/.m2/repository:/m2 \
   -v $HOME/.che/data:/data \
   --user=1000 \
   eclipse/che:nightly \
    archetype  generate --fast \
   --archid=che-plugin-ide-menu-archetype \
   --archversion=5.4.0-SNAPSHOT \
   --group=ofef.copm \
   --version=1.0-SNAPSHOT \
   --id=com-wizard \
   --no:interactive
```
but I have 
```
INFO: Welcome to Eclipse Che!
INFO:
INFO: You are missing a mandatory parameter:
INFO:    1. Mount 'docker.sock' for accessing Docker with unix sockets.
INFO:    2. Or, set DOCKER_HOST to Docker's location (unix or tcp).
INFO:
INFO: Mount Syntax:
INFO:    Start with 'docker run -it --rm -v /var/run/docker.sock:/var/run/docker.sock' ...
INFO:
INFO: DOCKER_HOST Syntax:
INFO:    Start with 'docker run -it --rm -e DOCKER_HOST=<daemon-location> ...'
INFO:
INFO: Possible root causes:
INFO:    1. Your admin has not granted permissions to /var/run/docker.sock.
INFO:    2. You passed '--user uid:gid' with bad values.
INFO:    3. Your firewall is blocking TCP ports for accessing Docker daemon.
```

benoitf commented:
@skabashnyuk is that your user id is 1000 ?

skabashnyuk commented:
yes. from /etc/password
```
sj:x:1000:1000:sj,,,:/home/sj:/bin/bash
```

benoitf commented:
@skabashnyuk in your command line I don't see
```
-v /etc/group:/etc/group:ro -v /etc/passwd:/etc/passwd:ro 
```

skabashnyuk commented:
Not helped
```
[sj@sj-i7:dockerfiles][cli-archetype]$ docker run  -it --rm  \
>    -v /var/run/docker.sock:/var/run/docker.sock  \
>    -v $HOME/dev/tmp/arch:/archetype \
>    -v $HOME/hdd/.m2/repository:/m2 \
>    -v $HOME/.che/data:/data \
>    -v /etc/group:/etc/group:ro \
>    -v /etc/passwd:/etc/passwd:ro \
>    --user=1000 \
>    eclipse/che:nightly \
>     archetype  generate --fast \
>    --archid=che-plugin-ide-menu-archetype \
>    --archversion=5.4.0-SNAPSHOT \
>    --group=ofef.copm \
>    --version=1.0-SNAPSHOT \
>    --id=com-wizard \
>    --no:interactive
INFO: Welcome to Eclipse Che!
INFO:
INFO: You are missing a mandatory parameter:
INFO:    1. Mount 'docker.sock' for accessing Docker with unix sockets.
INFO:    2. Or, set DOCKER_HOST to Docker's location (unix or tcp).
INFO:
```
Also I thought it suppose to be added in next docker command from cli 
https://github.com/eclipse/che/pull/4179/files#diff-50d5814c049c15267eb8acd7e446d1beR72

mkuznyetsov commented:
--che & --codenvy flags for build, run and stop commands.
As part of https://github.com/eclipse/che/issues/4484.
5a1a5dd
8550165

benoitf commented:
@skabashnyuk 
Here is a command to keep permissions on Linux
```
-v /etc/group:/etc/group:ro -v /etc/passwd:/etc/passwd:ro --user=1000:100 --group-add docker
```

1000 is my local uid
100 is my local gid

I add extra group docker so my user inside the container has two groups : 100 + docker



Full command should look like:
```
docker run  -it --rm -v /var/run/docker.sock:/var/run/docker.sock -v $HOME/dev/tmp/arch:/home/user/che-build -v $HOME/.m2:/home/user/.m2 -v /etc/group:/etc/group:ro -v /etc/passwd:/etc/passwd:ro -w /home/user/che-build --user=1000:100 --group-add 999 eclipse/che-dev:nightly mvn -Dmaven.repo.local=/home/user/.m2/repository/ org.apache.maven.plugins:maven-archetype-plugin:2.4:generate -DarchetypeRepository=http://maven.codenvycorp.com/content/groups/public/ -DarchetypeGroupId=org.eclipse.che.archetype -DarchetypeArtifactId=che-plugin-ide-menu-archetype -DarchetypeVersion=5.4.0-SNAPSHOT -DgroupId=ofef.copm -DartifactId=com-wizard -Dversion=1.0.0-SNAPSHOT
```

skabashnyuk commented:
@TylerJewell @benoitf I can't figure out correct command. I doubt that end users will know what to do . Do we have plan  ***B***?

```
[sj@sj-i7:eclipse]$ docker run  -it --rm  \
>     -v /var/run/docker.sock:/var/run/docker.sock  \
>     -v $HOME/dev/tmp/arch-new:/archetype \
>     -v $HOME/hdd/.m2/repository:/m2 \
>     -v $HOME/.che/data:/data \
>     -v /etc/group:/etc/group:ro \
>     -v /etc/passwd:/etc/passwd:ro \
>     -v $HOME/dev/src/codenvy/che/dockerfiles/base/scripts/base:/scripts/base \
>     --user=1000:100  \
>     --group-add 999 \
>     eclipse/che-cli:nightly  \
>      archetype  generate --fast \
>     --archid=plugin-json-archetype \
>     --archversion=5.6.0-SNAPSHOT \
>     --group=ofef.copm \
>     --version=1.0-SNAPSHOT \
>     --id=com-wizard \
>     --no:interactive
WARNING: No swap limit support
/scripts/base/startup_02_pre_docker.sh: line 14: /data/cli.log: Permission denied
[sj@sj-i7:eclipse]$ docker run  -it --rm      -v /var/run/docker.sock:/var/run/docker.sock      -v $HOME/dev/tmp/arch-new:/archetype     -v $HOME/hdd/.m2/repository:/m2     -v $HOME/.che/data:/data     -v /etc/group:/etc/group:ro     -v /etc/passwd:/etc/passwd:ro     -v $HOME/dev/src/codenvy/che/dockerfiles/base/scripts/base:/scripts/base     --user=1000:100      --group-add 999     eclipse/che-cli:nightly       archetype  generate --fast     --archid=plugin-json-archetype     --archversion=5.6.0-SNAPSHOT     --group=ofef.copm     --version=1.0-SNAPSHOT     --id=com-wizard     --no:interactive^C
[sj@sj-i7:eclipse]$ ^C
[sj@sj-i7:eclipse]$ docker run  -it --rm  \
>     -v /var/run/docker.sock:/var/run/docker.sock  \
>     -v $HOME/dev/tmp/arch-new:/archetype \
>     -v $HOME/hdd/.m2/repository:/m2 \
>     -v $HOME/.che/data:/data \
>     -v /etc/group:/etc/group:ro \
>     -v /etc/passwd:/etc/passwd:ro \
>     -v $HOME/dev/src/codenvy/che/dockerfiles/base/scripts/base:/scripts/base \
>     --user=1000:1000  \
>     --group-add 999 \
>     eclipse/che-cli:nightly  \
>      archetype  generate --fast \
>     --archid=plugin-json-archetype \
>     --archversion=5.6.0-SNAPSHOT \
>     --group=ofef.copm \
>     --version=1.0-SNAPSHOT \
>     --id=com-wizard \
>     --no:interactive
WARNING: No swap limit support
/scripts/base/startup_02_pre_docker.sh: line 14: /data/cli.log: Permission denied
```

benoitf commented:
@skabashnyuk in your case you gave these arguments to the cli and AFAIK all --group-add or user that you gave to the first command are not propagated to the subcalls to the maven archetype command

You need to add these flags to the commands added for the archetypes.

TylerJewell commented:
We have utility functions within the CLI that will let you know if the
yunderlying OS is Linux / Windows, so we just need some branching logic
added there and then applied to the archetype.


Tyler Jewell | CEO | tyler@​codenvy.​com | 9​78​.8​84​.53​55

On Thu, Mar 30, 2017 at 6:06 AM, Florent BENOIT <notifications@github.com>
wrote:

> @skabashnyuk <https://github.com/skabashnyuk> in your case you gave these
> arguments to the cli and AFAIK all --group-add or user that you gave to the
> first command are not propagated to the subcalls to the maven archetype
> command
>
> You need to add these flags to the commands added for the archetypes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/che/pull/4179#issuecomment-290404894>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAX9CrJA-LUpQ2e7iswldxHpp2QCGRAqks5rq6jqgaJpZM4MFgAS>
> .
>

skabashnyuk commented:
>You need to add these flags to the commands added for the archetypes.
Honestly I have no idea what does it mean.

skabashnyuk commented:
@benoitf Can you fix this code to make it working?

benoitf commented:
@skabashnyuk I was understanding that it was your team that should implement while I was providing guidance in order to make "knowledge transfer purposes" but I may be wrong ?

skabashnyuk commented:
My team was working on archetypes. And yes we can add new archetypes to cli. That's quite easy. But I doubt that in reasonable time we can solve issue with Linux permission. That's why I've asked you to finalize Tyler's PR. Because you know what is going on here. We don't.

benoitf commented:
@skabashnyuk ok I've added propagation of groups into the branch.
You should try again your command

note as well that all directories used from the command need to be existing previously to match the correct permissions

in your case
```
$HOME/dev/tmp/arch-new
$HOME/hdd/.m2/repository
$HOME/.che/data
```
should have user permissions

skabashnyuk commented:
$HOME/dev/tmp/arch-new - removed
$HOME/hdd/.m2/repository - has correct permissions
```
[sj@sj-i7:.che]$ sudo rm -rf $HOME/.che
[sj@sj-i7:.che]$ docker run  -it --rm  \
>      -v /var/run/docker.sock:/var/run/docker.sock  \
>      -v $HOME/dev/tmp/arch-new:/archetype \
>      -v $HOME/hdd/.m2/repository:/m2 \
>      -v $HOME/.che/data:/data \
>      -v /etc/group:/etc/group:ro \
>      -v /etc/passwd:/etc/passwd:ro \
>      -v $HOME/dev/src/codenvy/che/dockerfiles/base/scripts/base:/scripts/base \
>      --user=1000:1000  \
>      --group-add docker \
>      eclipse/che-cli:nightly  \
>       archetype  generate --fast \
>      --archid=plugin-json-archetype \
>      --archversion=5.7.0-SNAPSHOT \
>      --group=ofef.copm \
>      --version=1.0-SNAPSHOT \
>      --id=com-wizard \
>      --no:interactive
docker: Error response from daemon: linux spec user: Unable to find group docker.
[sj@sj-i7:.che]$
[sj@sj-i7:.che]$
[sj@sj-i7:.che]$ docker run  -it --rm  \
>      -v /var/run/docker.sock:/var/run/docker.sock  \
>      -v $HOME/dev/tmp/arch-new:/archetype \
>      -v $HOME/hdd/.m2/repository:/m2 \
>      -v $HOME/.che/data:/data \
>      -v /etc/group:/etc/group:ro \
>      -v /etc/passwd:/etc/passwd:ro \
>      -v $HOME/dev/src/codenvy/che/dockerfiles/base/scripts/base:/scripts/base \
>      --user=1000:1000  \
>      --group-add 999 \
>      eclipse/che-cli:nightly  \
>       archetype  generate --fast \
>      --archid=plugin-json-archetype \
>      --archversion=5.7.0-SNAPSHOT \
>      --group=ofef.copm \
>      --version=1.0-SNAPSHOT \
>      --id=com-wizard \
>      --no:interactive
WARNING: No swap limit support
/scripts/base/startup_02_pre_docker.sh: line 14: /data/cli.log: Permission denied
[sj@sj-i7:.che]$
```
I do have you latest changes locally
```
* ab0be2484 2017-03-31 | propagate groups of the command line of eclipse che cli to the sub commands (HEAD -> cli-archetype, origin/cli-archetype) [Florent BENOIT]
```

benoitf commented:
@skabashnyuk 
what is your user id and group id ?

could you paste result of the command `$ id `

because 1000:100 is on my debian system with docker group being 999 but I don't know what are ids on your system

on my system:
```
uid=1000(florent) gid=100(users) groupes=100(users),24(cdrom),25(floppy),29(audio),30(dip),44(video),46(plugdev),108(netdev),999(docker)
```

benoitf commented:
@skabashnyuk 
also please paste result of command
```
ls -la $HOME/dev/tmp/arch-new
ls -la $HOME/hdd/.m2/repository
ls -la $HOME/.che/data
```
to be sure that it's not using root permissions

skabashnyuk commented:
```
[sj@sj-i7:.che]$ id
uid=1000(sj) gid=1000(sj) groups=1000(sj),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),113(lpadmin),128(sambashare),999(docker)
```
```
[sj@sj-i7:.che]$ ls -la $HOME/dev/tmp/arch-new
ls: cannot access '/home/sj/dev/tmp/arch-new': No such file or directory
[sj@sj-i7:.che]$ ls -la $HOME/dev/tmp/
total 8
drwxrwxr-x 2 sj sj 4096 бер 31 15:59 .
drwxrwxr-x 6 sj sj 4096 лют 27 10:52 ..
[sj@sj-i7:.che]$ ls -la $HOME/.che/data
ls: cannot access '/home/sj/.che/data': No such file or directory
[sj@sj-i7:.che]$ ls -la $HOME/.che/
ls: cannot access '/home/sj/.che/': No such file or directory
[sj@sj-i7:.che]$
```

```
[sj@sj-i7:.che]$ ls -la $HOME/hdd/.m2/repository
total 248
drwxrwxr-x 62 sj sj 4096 лют 28 14:19 .
drwxrwxr-x  3 sj sj 4096 лют  9 16:19 ..
drwxrwxr-x  3 sj sj 4096 лют  7 11:10 ant
drwxrwxr-x  3 sj sj 4096 січ 20 13:02 antlr
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 aopalliance
drwxrwxr-x  3 sj sj 4096 лют  6 10:27 archetype
drwxrwxr-x  8 sj sj 4096 лют  3 12:05 asm
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 avalon-framework
drwxrwxr-x  3 sj sj 4096 січ 20 13:01 backport-util-concurrent
drwxrwxr-x  4 sj sj 4096 лют 18 15:08 cglib
drwxrwxr-x  4 sj sj 4096 січ 31 16:54 ch
drwxrwxr-x  3 sj sj 4096 січ 20 13:01 classworlds
drwxrwxr-x  3 sj sj 4096 лют  7 11:10 colt
drwxrwxr-x 35 sj sj 4096 лют 19 15:15 com
drwxrwxr-x  4 sj sj 4096 січ 23 09:11 commons-beanutils
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 commons-chain
drwxrwxr-x  3 sj sj 4096 січ 20 13:01 commons-cli
drwxrwxr-x  3 sj sj 4096 січ 20 12:58 commons-codec
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 commons-collections
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 commons-digester
drwxrwxr-x  3 sj sj 4096 січ 20 13:05 commons-fileupload
drwxrwxr-x  3 sj sj 4096 січ 20 12:58 commons-httpclient
drwxrwxr-x  3 sj sj 4096 січ 20 12:58 commons-io
drwxrwxr-x  3 sj sj 4096 січ 20 12:58 commons-lang
drwxrwxr-x  4 sj sj 4096 січ 20 13:02 commons-logging
drwxrwxr-x  3 sj sj 4096 лют  9 16:38 commons-pool
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 commons-validator
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 dom4j
drwxrwxr-x  4 sj sj 4096 січ 31 16:54 doxia
drwxrwxr-x  3 sj sj 4096 січ 23 22:23 gr
drwxrwxr-x  6 sj sj 4096 січ 30 13:56 io
drwxrwxr-x  3 sj sj 4096 січ 31 16:53 jakarta-regexp
drwxrwxr-x 11 sj sj 4096 січ 20 13:07 javax
drwxrwxr-x  3 sj sj 4096 січ 20 13:04 jaxen
drwxrwxr-x  3 sj sj 4096 січ 20 13:09 jdom
drwxrwxr-x  3 sj sj 4096 січ 20 13:04 jline
drwxrwxr-x  3 sj sj 4096 січ 20 13:06 joda-time
drwxrwxr-x  3 sj sj 4096 січ 23 22:23 jtidy
drwxrwxr-x  3 sj sj 4096 січ 20 13:01 junit
drwxrwxr-x  3 sj sj 4096 січ 20 13:02 log4j
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 logkit
drwxr-xr-x  4 sj sj 4096 бер 24 09:22 my
drwxrwxr-x  7 sj sj 4096 лют  3 12:05 net
drwxrwxr-x  3 sj sj 4096 лют 28 14:19 ognl
drwxrwxr-x 58 sj sj 4096 бер  1 09:25 org
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 oro
drwxrwxr-x  5 sj sj 4096 січ 31 16:54 plexus
drwxrwxr-x  3 sj sj 4096 січ 20 13:02 regexp
drwxrwxr-x  3 sj sj 4096 січ 31 16:55 rome
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 sslext
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 stax
drwxrwxr-x  3 sj sj 4096 лют  7 11:10 tapestry
drwxrwxr-x  3 sj sj 4096 лют  9 16:38 testParent
drwxrwxr-x  3 sj sj 4096 січ 31 16:54 wsdl4j
drwxrwxr-x  4 sj sj 4096 лют  7 11:10 xalan
drwxrwxr-x  4 sj sj 4096 січ 20 13:09 xerces
drwxrwxr-x  3 sj sj 4096 січ 20 13:03 xml-apis
drwxrwxr-x  3 sj sj 4096 січ 20 13:07 xmlpull
drwxrwxr-x  3 sj sj 4096 лют 19 15:16 xml-resolver
drwxrwxr-x  3 sj sj 4096 січ 20 13:09 xmlunit
drwxrwxr-x  3 sj sj 4096 січ 20 13:09 xom
drwxrwxr-x  4 sj sj 4096 січ 23 09:14 xpp3
```

benoitf commented:
ok, so, could you create first the mounted folder
```
mkdir -p $HOME/dev/tmp/arch-new
mkdir -p $HOME/dev/tmp/
```

as volume mounted folder needs to be created prior to docker run if you want to set correct owner.

skabashnyuk commented:
```
[sj@sj-i7:.che]$ mkdir -p $HOME/dev/tmp/arch-new
[sj@sj-i7:.che]$ mkdir -p $HOME/dev/tmp/
[sj@sj-i7:.che]$ docker run  -it --rm  \
>      -v /var/run/docker.sock:/var/run/docker.sock  \
>      -v $HOME/dev/tmp/arch-new:/archetype \
>      -v $HOME/hdd/.m2/repository:/m2 \
>      -v $HOME/.che/data:/data \
>      -v /etc/group:/etc/group:ro \
>      -v /etc/passwd:/etc/passwd:ro \
>      -v $HOME/dev/src/codenvy/che/dockerfiles/base/scripts/base:/scripts/base \
>      --user=1000:1000  \
>      --group-add 999 \
>      eclipse/che-cli:nightly  \
>       archetype  generate --fast \
>      --archid=plugin-json-archetype \
>      --archversion=5.7.0-SNAPSHOT \
>      --group=ofef.copm \
>      --version=1.0-SNAPSHOT \
>      --id=com-wizard \
>      --no:interactive
WARNING: No swap limit support
/scripts/base/startup_02_pre_docker.sh: line 14: /data/cli.log: Permission denied
```

benoitf commented:
@skabashnyuk 

What is permissions for folder 
```
>      -v $HOME/.che/data:/data \
```

```
docker run  -it --rm       -v /var/run/docker.sock:/var/run/docker.sock       -v $HOME/arch-n:/archetype      -v $HOME/.m2/repository:/m2      -v $HOME/data:/data      -v /etc/group:/etc/group:ro      -v /etc/passwd:/etc/passwd:ro      --user=1000:1000 --group-add 999 --entrypoint=/bin/bash  eclipse/che-cli:nightly -c "ls -la /"
```

could you check if the folder is not created with root access
or delete first the folder, and recreate it under your sj user

```
$ sudo rm -rf $HOME/.che/data
$ mkdir -p $HOME/.che/data
```

skabashnyuk commented:
yes. It's working. I assume cause was wrong permissions $HOME/.che
Because I've used it without
```
     --user=1000:1000  \
     --group-add 999 \
```
to run che. Thx @benoitf

TylerJewell commented:
Is this a documentation issue that we need to add into the description of how the archetype command runs?  Or perhaps in the command help  that is printed to the console?  If so, someone please add a commit to the archetype cli pr.

Tyler

> On Mar 31, 2017, at 10:25 AM, Sergii Kabashniuk <notifications@github.com> wrote:
> 
> yes. It's working. I assume cause was wrong permissions $HOME/.che
> Because I've used it without
> 
>      --user=1000:1000  \
>      --group-add 999 \
> to run che. Thx @benoitf
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

TylerJewell commented:
The syntax is not really too complicated - because it's a one time usage
sort of thing for advanced devs.

And if you are a dev that is REALLY advanced, then the docs should just
have you use the mvn archetype:generate syntax natively anyways.


Tyler Jewell | CEO | tyler@​codenvy.​com | 9​78​.8​84​.53​55

On Wed, Apr 5, 2017 at 4:46 AM, Sergii Kabashniuk <notifications@github.com>
wrote:

> *@skabashnyuk* approved this pull request.
>
> This pr is doing that was declared. One thing that worries me is that
> syntax is too complicated.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/che/pull/4179#pullrequestreview-31024700>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAX9Cu57imDBzsRvgVMkK84OjnvJsD78ks5rs376gaJpZM4MFgAS>
> .
>

garagatyi commented:
I would suggest to provide an example on how to get user id in command line. E.g. instead of `--user={uid}` `--user=$(id -u)`

garagatyi commented:
Consider usage more clear names for scripts.
Instead of `cmd_abuild.sh` you can use `cmd_arch_build.sh`. Current names are not clear enough.

garagatyi commented on dockerfiles/base/scripts/base/commands/cmd_abuild.sh:
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+# Copyright (c) 2012-2016 Codenvy, S.A.
```
Can you change the year in all files to 2017

TylerJewell commented:
"arch" is not clear - it means architecture to most people.


Tyler Jewell | CEO | tyler@​codenvy.​com | 9​78​.8​84​.53​55

On Wed, Apr 5, 2017 at 5:13 AM, Alexander Garagatyi <
notifications@github.com> wrote:

> Consider usage more clear names for scripts.
> Instead of cmd_abuild.sh you can use cmd_arch_build.sh. Current names are
> not clear enough.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/che/pull/4179#issuecomment-291842339>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAX9CmALi7H7jnynDT2fKtaWPSw0dj4Cks5rs4VRgaJpZM4MFgAS>
> .
>

garagatyi commented:
Then use archetype. `arch` can't be less clear then `a`

garagatyi commented on dockerfiles/base/scripts/base/commands/cmd_archetype.sh:
```diff
@@ -0,0 +1,185 @@
+    ;;
+    all)
+      cmd_lifecycle agenerate || true
+      cmd_lifecycle abuild || true
```
Can you explain why we should build if archetype generation failed?
